[Webhook] update doc (#19974)
symfony/symfony-docs: webhook.rst

    The Webhook component was introduced in Symfony 6.3.

Essentially, webhooks serve as event notification mechanisms, typically via HTTP POST requests, enabling real-time updates.

> The Webhook component is used to respond to remote webhooks in order to trigger actions
> in your application. Additionally, it can assist in dispatching webhooks from the provider side.

Reviewer: ```suggestion
The Webhook component has two primary functions:
1. Consumer: Listen and respond to remote webhooks dispatched by 3rd party services.
2. Provider: Dispatch webhooks to 3rd party services.
```

Author: indeed, it might be clearer to present it this way

    $ composer require symfony/webhook

Consuming Webhooks
------------------

Consider an example of an API where it's possible to track the stock levels of various products.
A webhook has been registered to notify when certain events occur, such as stock depletion for a specific product.

During the registration of this webhook, several pieces of information were included in the POST request,
including the endpoint to be called upon the occurrence of an event, such as stock depletion for a certain product:

.. code-block:: json
    {
      "name": "a name",
      "url": "something/webhook/routing_name"
      "signature": "..."
      "events": ["out_of_stock_event"]
      ....
    }

From the perspective of the consumer application, which receives the webhook, three primary phases need to be anticipated:

1) Receiving the webhook

2) Verifying the webhook and constructing the corresponding Remote Event

3) Manipulating the received data.

> Symfony Webhook, when used alongside Symfony Remote Event, streamlines the management of these fundamental phases.

Author: todo: update RemoteEvent

    $ composer require symfony/webhook

Consuming Webhooks
------------------

Consider an example of an API where it's possible to track the stock levels of various products.

> A webhook has been registered to notify when certain events occur, such as stock depletion for a specific product.

Reviewer: This is a 3rd party service that can send webhooks to your Symfony app? This should be clarified.

> During the registration of this webhook, several pieces of information were included in the POST request,

Reviewer: This is information that is typically available in the 3rd party service documentation?

including the endpoint to be called upon the occurrence of an event, such as stock depletion for a certain product:

.. code-block:: json

    {
      "name": "a name",
      "url": "something/webhook/routing_name",
      "signature": "...",
      "events": ["out_of_stock_event"],
      ....
    }

> From the perspective of the consumer application, which receives the webhook, three primary phases need to be anticipated:

Reviewer: ```suggestion
From the perspective of your application (the *consumer*), which receives the webhook, three primary phases need to be anticipated:
```

1) Receiving the webhook

2) Verifying the webhook and constructing the corresponding Remote Event

> 3) Manipulating the received data.

Reviewer: Manipulating or "using"?

Author: Indeed, I would rather use "using"

    $ composer require symfony/webhook

Consuming Webhooks
------------------

Consider a third-party API that allows you to track stock levels for various products and can send webhooks to your
Symfony application.

From the perspective of your application (the *consumer*), which receives the webhook, three primary phases need to be anticipated:

1) Receiving the webhook

2) Verifying the webhook and constructing the corresponding Remote Event

3) Using the received data.

Symfony Webhook, when used alongside Symfony RemoteEvent, streamlines the management of these fundamental phases.

A Single Entry Endpoint: Receive

> --------------------------------

Reviewer 2: shouldn't this be a sub-heading of the `Consuming Webhooks` section instead ? With the new doc covering both consuming and providing webhooks, I think we should have only 3 `h2` titles: `Installation`, `Consuming webhooks` and `Providing webhooks`

Through the built-in :class:`Symfony\\Component\\Webhook\\Controller\\WebhookController`, a unique entry point is offered to manage all webhooks
that our application may receive, whether from the Twilio API, a custom API, or other sources.

By default, any URL prefixed with ``/webhook`` will be routed to this :class:`Symfony\\Component\\Webhook\\Controller\\WebhookController`.
Additionally, you have the flexibility to customize this URL prefix and rename it according to your preferences.

.. code-block:: yaml

    # config/routes/webhook.yaml
    webhook:
        resource: '@FrameworkBundle/Resources/config/routing/webhook.xml'
        prefix: /webhook # or possible to customize

Additionally, you must specify the parser service responsible for analyzing and parsing incoming webhooks.
It's crucial to understand that the :class:`Symfony\\Component\\Webhook\\Controller\\WebhookController` itself remains provider-agnostic, utilizing
a routing mechanism to determine which parser should handle incoming webhooks for analysis.

> As mentioned earlier, incoming webhooks require a specific prefix to be directed to the :class:`Symfony\\Component\\Webhook\\Controller\\WebhookController`.

Reviewer 2: I think this section gives far too much details about the WebhookController, which are irrelevant for most users as they won't ever touch it directly, while not explaining the high-level picture:

> webhooks are received on a URL `<prefix>/<routing-name>` and routed to the appropriate request parser and handler based on the routing name.

That's the info that users need to have, with configuration examples in all formats for the routing configuration

This prefix forms the initial part of the URL following the domain name.
The subsequent part of the URL, following this prefix, should correspond to the routing name chosen in your configuration.

The routing name must be unique as this is what connects the provider with your
webhook consumer code.

> .. code-block:: yaml

Reviewer 2: this should use a `configuration-block` to show the configuration in all formats

If you used the command to set this up, it was created automatically

.. code-block:: terminal

    $ php bin/console make:webhook

Otherwise, you'll need to manually add it with the ``AsRemoteEventConsumer`` attribute which will allow you to designate this class as a consumer implementing :class:`Symfony\\Component\\RemoteEvent\\Consumer\\ConsumerInterface`,
making it recognizable to the RemoteEvent component so it can pass the converted object to it.
Additionally, the name passed to your attribute is critical; it must match the configuration entry under routing that you specified in the ``webhook.yaml`` file, which in your case is ``my_first_parser``.

In the :method:`Symfony\\Component\\RemoteEvent\\Consumer\\ConsumerInterface::consume` method,
you can access your object containing the event data that triggered the webhook, allowing you to respond appropriately.

For example, you can use Mercure to broadcast updates to clients of the hub, among other actions ...::

    // src/Webhook/ExampleRequestParser.php

>     #[AsRemoteEventConsumer('my_first_parser')] # routing name

Reviewer 2: ```suggestion
    #[AsRemoteEventConsumer('my_first_parser')] // routing name
```

We don't use `#` as a delimiter for line comments in the Symfony coding standards

    final class ExampleWebhookConsumer implements ConsumerInterface
    {
        public function __construct()
        {
        }

        public function consume(RemoteEvent $event): void
        {
            // Implement your own logic here
        }
    }

If you are using it alongside other components that already include built-in parsers,
you will need to configure the settings (as mentioned earlier) and also create your own consumer.
This is necessary because it involves your own business logic and your specific reactions to the remote event(s) that may be received from the built-in parsers.

> Usage in Combination with the Mailer Component

Reviewer 2: I'm not sure having 2 sections about using it in combination with mailer and notifier in the page (one in the section about the parser and one in the section about the handler) makes sense. I think it might be better to organize the content the other way around:

- usage with mailer
- usage with notifier
- usage with custom webhooks (talking about both the request parser and the handler)

    final class ExamplePayloadConverter implements PayloadConverterInterface
    {
        public function convert(array $payload): AbstractPriorityStockRefillEvent
        {
            ...

            if (....) {
                $event = new HighPriorityStockRefillEvent($name, $payload['id]', $payload])
            } elseif {
                $event = new MediumPriorityStockRefillEvent($name, $payload['id]', $payload])
            } else {
                $event = new LowPriorityStockRefillEvent($name, $payload['id]', $payload])
            }

            ....

            return $event;
        }
    }

From this, we can see that the Remote Event component is highly beneficial for handling webhooks.
It enables you to convert the incoming webhook data into validated objects that can be efficiently manipulated and utilized according to your requirements.

Remote Event Consumer: Handling and Manipulating The Received Data
------------------------------------------------------------------

It is important to note that when the incoming webhook is processed by the :class:`WebhookController`, you have the option to handle the consumption of remote events asynchronously.

> Indeed, this can be configured using a bus, with the default setting pointing to the Messenger component's default bus.

Author: todo: show config

        }
    }

Providing Webhooks
------------------

Symfony Webhook and Symfony Remote Event, when combined with Symfony Messenger, are also useful for APIs responsible for dispatching webhooks.

For instance, you can utilize the specific :class:`SendWebhookMessage` and :class:`SendWebhookHandler` provided to dispatch the message either synchronously or asynchronously using the Symfony Messenger component.

The SendWebhookMessage takes a :class:`Subscriber` as its first argument, which includes the destination URL and the mandatory secret.
If the secret is missing, an exception will be thrown.

> As a second argument, it expects a :class:`RemoteEvent` containing the webhook name, the ID, and the payload, which is the substantial information you wish to communicate.

Author: I need to elaborate a little bit more

The :class:`SendWebhookHandler` configures the headers, the body of the request, and finally sign the headers before making an HTTP request to the specified URL using Symfony's HttpClient component::

    $subscriber = new Subscriber($urlCallback, $secret);

    $event = new Event(‘name.event, ‘1’, […]);

    $this->bus->dispatch(new SendWebhookMessage($subscriber, $event));

> 

Author: add for a customised approach using the `Symfony\Component\Webhook\Server\TransportInterface`
